docs_new/features/core/output_types.md

```diff
@@ -0,0 +1,173 @@
+output_type = PizzaOrBurger
+```
+
+### Structured Objects
+
+What we call structured object here are roughly all common types that can be turned into a JSON schema. That includes:
+- A Pydantic class
+- A Dataclass
+- A TypedDict
+- A GenSon SchemaBuilder
+- A Callable (the parameters are turned into the keys and the type hinting is used to define the types of the values)
+
+For instance:
+
+```python
+from dataclasses import dataclass
+
+@dataclass
+class Character:
+    name: str
+    age: int
+
+ouput_type = Character
+
+def character(name: str, age: int):
+    return None
+
+output_type = character
+```
+
+On top of those five types that can directly be used with a model, there are also two other json schema formats for which we need to use an Outlines-specific class: JSON schema strings and dicts. As those do not have a distinct type, we must wrap them into an `outlines.types.JsonSchema` object. For instance:
+
+```python
+from outlines.types import JsonSchema
+
+schema_string = '{"type": "object", "properties": {"answer": {"type": "number"}}}'
+output_type = JsonSchema(schema_string)
+
+schema_dict = {
+    "type": "object",
+    "properties": {
+        "answer": {"type": "number"}
+    }
+}
+output_type = JsonSchema(schema_dict)
+```
+
+### Regex Patterns
+
+You can ask your model to generate text that follows a regex pattern. As the latter are expressed as simple raw string literals, we need to use an Outlines-specific class. Thus, the regex string should be be wrapped in an `outlines.types.Regex` object. For instance:
+
+```python
+from outlines.types import Regex
+
+regex = r"[0-9]{3}"
+output_type = Regex(regex)
+```
+
+The `types` module contains a few common regex patterns stored in variable you can import and directly use as output types. There are for instance the patterns for a sentence, an email address and an isbn reference. For instance:
+
+```python
+from outlines.types import sentence
+
+print(type(sentence)) # outlines.types.dsl.Regex
+print(sentence.pattern) # [A-Z].*\s*[.!?]
+```
+
+To help you create complex regex patterns yourself, you can rely on the Outlines [regex DSL]().
+
+### Context-Free Grammars
+
+Outlines allows you to generate text that respects the syntax of a context-free grammar. For that, you need to define a Lark grammar. As the latter is expressed as a string, we need to use an Outlines-specific class. Thus, the CFG string should be be wrapped in an `outlines.types.CFG` object. For instance:
+
+```python
+from outlines.types import CFG
+
+grammar_string = """
+    start: expr
+    expr: "{" expr "}" | "[" expr "]" |
+"""
+output_type = CFG(grammar_string)
+```
+
+You can find a few Lark grammar examples in the [grammars module]().
+
+## Output type availability
+
+The output types presented above are not available for all models as some have only limited support for structured outputs. Please refer to the documentation of the specific model you wish to use to know what output types it supports.
```

Comment:
We should probably add a `models/index.md` file in the root of the `models/` dir, and then update the mkdocs tree to point to that index file. See how `index.md` is included directly under the name [here](https://github.com/dottxt-ai/outlines/blob/b6ffccc5581aa920e2af76a8d10ea8345ce86df6/mkdocs.yml#L149-L153)).

That index file should contain hyperlinks to each of the available models, and eventually, a table listing supported features. We can defer the table to a much later date.